rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,634 @@
+       *self.get_tool_metadata(name)?.component_id
+    }
+
+    /// Gets a reference to the specified modal dev tool by type name, as supplied by the `ModalDevTool` trait.
+    fn get_tool_by_name(&self, world: &World, name: &str) -> Option<&dyn ModalDevTool> {
+        let component_id = self.lookup_tool_component_id(name)?;
+        DevToolsRegistry::get_by_id(world, component_id)
+    }
+
+    /// Gets a mutable reference to the specified modal dev tool by name.
+    fn get_tool_mut_by_name(&self, mut world: &mut World, component_id: ComponentId) -> Option<&mut dyn ModalDevTool> {
+        let component_id = self.lookup_tool_component_id(name)?;
+        DevToolsRegistry::get_mut_by_id(world, component_id)
+    }
+    
+    /// Iterates over the list of registered modal dev tools.
+    fn iter_tools(&self, world: &World) -> impl Iterator<Item = (ComponentId, &dyn ModalDevTool)> { 
+        self.modal_dev_tools.iter().map(|&id| (id, self.get(world, id)))
+    }
+
+    /// Mutably iterates over the list of registered modal dev tools.
+    fn iter_tools_mut(&self, mut world: &mut World) -> impl Iterator<Item = (ComponentId, &mut dyn ModalDevTool)> { 
+        self.modal_dev_tools.iter_mut().map(|&id| (id, self.get(world, id)))
+    }
+
+    /// Iterates over the list registered dev commands, returning their name and `DevCommandMetadata`.
+    fn iter_commands(&self) -> impl Iterator<Item = (&str, `DevCommandMetadata)> {
+        self.dev_commands.iter()
+    }
+}
+```
+
+Once the user has a `dyn ModalDevTool`, they can perform whatever operations they'd like: enabling and disabling it, reading metadata and so on.
+There's no need to add further duplicate API: users building toolboxes are generally sophisticated, and the number of calls should be quite small.
+
+While we can use the [dynamic resource APIs](https://docs.rs/bevy/latest/bevy/ecs/world/struct.World.html#method.get_resource_by_id) and trait casting to get references to the list of modal dev tools, we cannot do the same for the dev commands!
+While this may seem unintuitive, the reason for this is fairly simple: modal configuration is always present, while the configuration for each dev command is generated when it is sent.
+As a result, we can only access its metadata: the arguments it takes, its doc strings and so on.
+
+### Parsing and inserting modal dev tools
+
+One method on `DevToolsRegistry` is worth special attention: `parse_and_insert_tool`.
+This tool takes a common pattern, parsing the configuration for a dev tool from a provided string, and provides a user friendly, safe API over it.
+
+```rust
+impl DevToolsRegistry {
+    /// Parses the given str `s` into a modal dev tool corresponding to its name if possible.
+    ///
+    /// For a typed equivalent, simply use the `FromStr` trait that this method relies on directly.
+    fn parse_and_insert_tool(&self, world: &mut World, s: &str) -> Result<(), DevToolParseError>{
```

Comment:
First: What are we parsing here? I was having a look and ModalDevTool trait doesn't have a metadata at all. If we expect to not only parse commands here, but also tools themselves, we can store in the Metadata some kind of identifier to it: Maybe a enum of some kind?

I Think we need to clarify how are we going to register commands so this can be an easy way forward. 
For example, lets say we store in the metadata an enum 
```rust
enum DevToolType {
   ModalDevTool,
   DevCommand
}
```
Now, something that maybe can be strange: To do everything the way we currently do with strings, we must ensure that all the tools have different names.
And so, we get the tool metadata, with the enum stored, and we can make a command or insert the tool, as simple as possible.

Reply:
> First: What are we parsing here? I was having a look and ModalDevTool trait doesn't have a metadata at all

This is an omission! Lemme fix that...